packages/contracts/contracts/engine/V3/EngineFactoryV0.sol

```diff
@@ -0,0 +1,349 @@
+            IEngineFactoryV0.EngineCoreType.Engine
+            ? engineImplementation
+            : engineFlexImplementation;
+
+        engineContract = Clones.cloneDeterministic({
+            implementation: implementation,
+            salt: salt
+        });
+
+        IGenArt721CoreContractV3_Engine(engineContract).initialize(
+            engineConfiguration,
+            adminACLContract
+        );
+
+        (
+            string memory coreContractType,
+            string memory coreContractVersion
+        ) = engineCoreType == IEngineFactoryV0.EngineCoreType.Engine
+                ? (coreType, coreVersion)
+                : (flexCoreType, flexCoreVersion);
+
+        // register the new Engine contract
+        ICoreRegistryV1(coreRegistry).registerContract(
+            engineContract,
+            stringToBytes32(coreContractVersion),
+            stringToBytes32(coreContractType)
+        );
+        // emit event
+        emit EngineContractCreated(engineContract);
+    }
+
+    /**
+     * @notice Registers multiple contracts with the core registry.
+     * @param contractAddresses An array of contract addresses to register.
+     * @param coreVersions An array of versions corresponding to the contract addresses.
+     * @param coreTypes An array of types corresponding to the contract addresses.
+     */
+    function registerMultipleContracts(
+        address[] calldata contractAddresses,
+        bytes32[] calldata coreVersions,
+        bytes32[] calldata coreTypes
+    ) external onlyOwner {
+        require(
+            contractAddresses.length == coreVersions.length &&
+                contractAddresses.length == coreTypes.length,
+            "Mismatched input lengths"
+        );
+
+        ICoreRegistryV1(coreRegistry).registerContracts(
+            contractAddresses,
+            coreVersions,
+            coreTypes
+        );
+    }
+
+    /**
+     * @notice Unregisters multiple contracts from the core registry.
+     * @param contractAddresses An array of contract addresses to unregister.
+     */
+    function unregisterMultipleContracts(
+        address[] calldata contractAddresses
+    ) external onlyOwner {
+        ICoreRegistryV1(coreRegistry).unregisterContracts(contractAddresses);
+    }
+
+    /**
+     * @notice Abandons the contract, preventing it from being used to create
+     * new Engine and Engine Flex contracts.
+     * Only callable by the owner, and only once; reverts otherwise.
+     */
+    function abandon() external onlyOwner {
+        require(!isAbandoned, "factory is abandoned");
+        // set isAbandoned to true
+        isAbandoned = true;
+        // emit event
+        emit Abandoned();
+    }
+
+    /**
+     * @dev This contract is not intended to hold funds. This function,
+     * `drainETH`, and `drainERC20` are implemented to prevent the loss
+     * of funds that might be sent to this contract inadvertently.
+     */
+    receive() external payable {}
```

Comment:
@ryley-o Added with a clarifying comment

Reply:
nice - was this function required?

Reply:
cool, I forgot that is required - from the docs
```
When Ether is sent directly to a contract (without a function call, i.e. sender uses send or transfer) but the receiving contract does not define a receive Ether function or a payable fallback function, an exception will be thrown, sending back the Ether (this was different before Solidity v0.4.0).
```